### PR TITLE
test(ICRC_ledger): FI-1542: Add fee collector test for icrc3_get_blocks

### PR DESCRIPTION
**Pull request by mbjorkqvist**
Verify that the `fee_collector` is set also in ICRC3 blocks, and that it is set not only for `transfer` transactions, but also `approve` and `transfer_from` transactions.

### CODE REVIEW

File: rs/ledger_suite/tests/sm-tests/src/fee_collector.rs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+                    ),
+                },
+                _ => panic!("Unexpected account representation: {:?}", array),
+            },
+            value => panic!("Expected Value::Array but found {:?}", value),
+        }
+    }
+
+    fn fee_collector_from_block(
+        block: icrc_ledger_types::icrc::generic_value::Value,
```
Comment (fsodre) [inline rs/ledger_suite/tests/sm-tests/src/fee_collector.rs:177]: Why isn't `block` an immutable reference here? It seems we don't need ownership but we end up cloning it every time we call this method.

File: rs/ledger_suite/tests/sm-tests/src/fee_collector.rs [hunk 2]
```diff
@@ -251,7 +267,16 @@ where
...
+                let blocks = match block_retrieval {
+                    BlockRetrieval::Legacy => {
+                        get_blocks(&env, ledger_id.get().0, block_id, 3).blocks
+                    }
+                    BlockRetrieval::Icrc3 => icrc3_get_blocks(&env, ledger_id, block_id, 3)
+                        .blocks
+                        .into_iter()
+                        .map(|b| GenericBlock::from(b.block))
+                        .collect(),
+                };
```
Comment (fsodre) [inline rs/ledger_suite/tests/sm-tests/src/fee_collector.rs]: Although not exactly the same, it seems like mostly what's going on in 226-233. Could we refactor this to decrease complexity in this function?

File: rs/ledger_suite/tests/sm-tests/src/fee_collector.rs [hunk 3]
```diff
@@ -199,15 +199,20 @@ pub fn test_fee_collector_blocks<T>(
...
                 1..10_000_000u64,
             )
-                .prop_filter("The three accounts must be different", |(a1, a2, a3, _)| {
-                    a1 != a2 && a2 != a3 && a1 != a3
-                }),
-            |(account_from, account_to, fee_collector_account, amount)| {
+                .prop_filter(
+                    "The three accounts must be different",
+                    |(a1, a2, a3, a4, _)| {
+                        a1 != a2 && a2 != a3 && a1 != a3 && a1 != a4 && a2 != a4 && a3 != a4
```
Comment (fsodre) [inline rs/ledger_suite/tests/sm-tests/src/fee_collector.rs:208]: This is a bit hard to manage (as in "visually checking it's correct and later make changes to it").
Could we do something like: `|(a1, a2, a3, a4, _)| HashSet::from([a1, a2, a3, a4]).len() == 4` ?

File: rs/ledger_suite/tests/sm-tests/src/fee_collector.rs (hunk 1 above)
Comment (mbjorkqvist) [reply rs/ledger_suite/tests/sm-tests/src/fee_collector.rs]: Good question! I just moved this code from the parent module, but you're right that there's a bunch of unnecessary cloning going on. I changed all the helper functions to take immutable references and removed a bunch of calls to `clone()`.

File: rs/ledger_suite/tests/sm-tests/src/fee_collector.rs (hunk 3 above)
Comment (mbjorkqvist) [reply rs/ledger_suite/tests/sm-tests/src/fee_collector.rs]: Great suggestion, thanks - fixed!

File: rs/ledger_suite/tests/sm-tests/src/fee_collector.rs (hunk 2 above)
Comment (mbjorkqvist) [reply rs/ledger_suite/tests/sm-tests/src/fee_collector.rs]: Good idea, done!

**Review by fsodre — APPROVED** [review]

**Comment by mbjorkqvist** [comment]
Thanks for the feedback @fsodre !

**Pull request merged**